File: LeadToOffer/51_ConstructMulMatrix.cpp
//1. O(n^2) time, O(1) space.
class Solution {
public:
    vector<int> multiply(const vector<int>& A) {
        vector<int> res;
        if (A.size()<=1)
            return res;
        int n = A.size();
        for (int i = 0; i < n; i++)
            res.push_back(1);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < n; j++) {
                res[j] *= A[ (j+i)%n ];
            }
        }
        return res;
    }
};


//2. Use two array to store the prod from left and from right. DP.
// O(n) space, O(n) time.
class Solution {
public:
    vector<int> multiply(const vector<int>& A) {
        vector<int> res;
        if (A.size()<=1)
            return res;
        int n = A.size();
        vector<int> left;
        vector<int> right;
        left.push_back(1);
        right.push_back(1);
        for (int i = 0; i < n-1; i++) {
            left.push_back(left[i]*A[i]);
            right.push_back(right[i]*A[n-i-1]);
        }
        for (int i = 0; i < n; i++) {
                res.push_back(left[i]*right[n-i-1]);
        }
        return res;
    }
};

File: Algo/LeadToOffer/51_ConstructMulMatrix.py
class Solution(object):
    def constructArr(self, a):
        """
        :type a: List[int]
        :rtype: List[int]
        """
        n = len(a)
        l, r = [1], [1]
        for i in range(0, n-1):
            l.append(l[i] * a[i])
            r.append(r[i] * a[n-1-i])

        res = []
        for i in range(0, n):
            res.append(l[i] * r[n - 1 - i])

        return res


# O(1) space
class Solution(object):
    def constructArr(self, a):
        """
        :type a: List[int]
        :rtype: List[int]
        """
        n = len(a)
        res = [1 for i in range(n)]
        for i in range(0, n-1):
            res[i+1] = res[i] * a[i]

        fac = 1
        for i in range(0, n):
            res[n-1-i] = res[n-1-i] * fac
            fac *= a[n-1-i]

        return res